### ISSUE TITLE
[FR]: Support JSON Transcoding for gRPC services -- allows services to be called via plain HTTP

### ISSUE THREAD

**Issue by GavinRay97** [quarkusio/quarkus#25837]
### Description

It would be really useful if Quarkus' gRPC implementation allowed configuring JSON Transcoding:
- https://cloud.google.com/endpoints/docs/grpc/transcoding
- https://devblogs.microsoft.com/dotnet/announcing-grpc-json-transcoding-for-dotnet/

(Maybe it already does this and I just missed the docs/tests?)

```proto
syntax = "proto3";
import "google/api/annotations.proto";
package greet;

service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {
      get: "/v1/greeter/{name}"
    };
  }
}

message HelloRequest {
  string name = 1;
}

message HelloReply {
  string message = 1;
}
```

**Comment by michalszynkiewicz** [comment]
Quarkus does not support it at the moment but it is a nice idea. I have no idea where we'll have time to implement it though, at this point.

**Comment by cescoffier** [comment]
Yes, it looks like a good plan. I would not register endpoints in RestEasy reactive but work at the 'router' level directly (so you do not need RestEasy reactive, just vertx htttp which is the base of everything Http related).

Also we need to think about streams. Did you think about them?

**Comment by hardikp-fens** [comment]
I wanted to utilize this feature in my quarkus project, so are there any updates regarding this?

/cc @cescoffier, @michalszynkiewicz

**Comment by schonbachler** [comment]
This would be a very useful feature. gRPC-web transcoding would be a related feature. As of now - a possible solution is to use a reverse-proxy (like envoy) in front, that does the transcoding.

**Comment by cescoffier** [comment]
Vert.x is coming with an implementation of the gRPC Web "proxy" which would be built in in the gRPC server.

CC @tsegismont

**Comment by cescoffier** [comment]
Vert.x 5 is coming with gRPC Web support - We will integrate gRPC Web at that point (2025).

**Comment by schonbachler** [comment]
@cescoffier Great news. Please keep in mind that grpc web and grpc JSON transcoding are two different things.

**Comment by cescoffier** [comment]
@schonbachler I know, but at the moment, our priority is gRPC Web. Contributions are more than welcome for gRPC JSON.

**Comment by ilyakharlamov** [comment]
This feature of `google.api.http` transcoding feature is supported by the following implementations:

* [Google APIs](https://github.com/googleapis/googleapis) & [Google Cloud](https://cloud.google.com/endpoints) (internal google implementation, closed source)
* [gRPC Gateway proxy](https://github.com/grpc-ecosystem/grpc-gateway) (Golang, open source)
* [Envoy proxy](https://github.com/envoyproxy/envoy) (C++, open source)
* [Microsoft's ASP.NET](https://learn.microsoft.com/en-us/aspnet/core/grpc/json-transcoding?view=aspnetcore-9.0) (C#, open source)

Microsoft's solution seems to be the most feature-complete also supporting Swagger/OpenAPI gen for the transcoded GRPC similar to [quarkus-smallrye-openapi](https://quarkus.io/extensions/io.quarkus/quarkus-smallrye-openapi/).

Even though it is relatively simple (as PROTO-to-JSON  as well as JSON-to-PROTO conversions are [standardized](https://protobuf.dev/programming-guides/json/) and it's literally a single method call), currently, no Java server supports this, so it could be beneficial for widespread adoption of Quarkus framework.

**Comment by cescoffier** [comment]
It won't happen before Quarkus 4 as it requires Vertx 5.

**Comment by krickert** [comment]
Not sure if this issue is a good place - I'd like to help on this.  I created this feature for micronaut last year, and it wasn't too bad to do.  It would be far easier to do with quarkus - 

I'm hoping to take this a step further and automatically register grpc descriptors to a schema registry too.  I'm doing this now with apicurio, but would recommend using the confluent API endpoints for consistency.

I bring this up because such descriptor integration can also lend to open API registration - so msft has open API, we quarkus folk could have open api   auto-generated from descriptors from a schema manager.

All this can be done with 0 reflection.

**Comment by cescoffier** [comment]
@krickert The plan was to use the Vert.x 5 feature. I haven't looked at it yet, but if you have the bandwidth, please take a look and tell us what's missing.

**Comment by zZHorizonZz** [comment]
We have added grpc transcoding a couple of months back to vertx 5 grpc. Recently I have opened a PR which extends the generator with openapi schema generation capability, we will see where that goes. Also I'm not quite sure if quarkus will be using the new generator or if it will do something on its own.
